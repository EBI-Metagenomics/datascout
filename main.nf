#!/usr/bin/env nextflow
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Ensembl/datascout
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Github : https://github.com/manuelcarbajo/datascout
    Github : https://github.com/Ensembl/datascout
----------------------------------------------------------------------------------------
*/

nextflow.enable.dsl = 2

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT FUNCTIONS / MODULES / SUBWORKFLOWS / WORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { DATASCOUT  } from './workflows/datascout'
include { PIPELINE_INITIALISATION } from './subworkflows/local/utils_nfcore_datascout_pipeline'
include { PIPELINE_COMPLETION     } from './subworkflows/local/utils_nfcore_datascout_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NAMED WORKFLOWS FOR PIPELINE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// WORKFLOW: Run main analysis pipeline depending on type of input
//
workflow ENSEMBL_DATASCOUT {

    take:
    csv_file // channel: samplesheet read in from --csv-file
    outdir
    orthodb_dir

    main:

    //
    // WORKFLOW: Run pipeline
    //
    DATASCOUT (
        csv_file,
        outdir,
        orthodb_dir
    )

    //emit:
    //datascout_report = DATASCOUT.out.datascout_report

}
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow {

    main:

    //
    // SUBWORKFLOW: Run initialisation tasks
    //
    PIPELINE_INITIALISATION (
        params.version,
        params.help,
        params.validate_params,
        params.monochrome_logs,
        args,
        params.outdir,
        params.csv_file
    )

    //
    // WORKFLOW: Run main workflow
    //
    ENSEMBL_DATASCOUT (
        PIPELINE_INITIALISATION.out.csv_file,
        PIPELINE_INITIALISATION.out.outdir,
        PIPELINE_INITIALISATION.out.orthodb_dir
    )

    //
    // SUBWORKFLOW: Run completion tasks
    //
    //PIPELINE_COMPLETION (
    //    params.outdir,
    //    params.monochrome_logs,
    //    ENSEMBL_DATASCOUT.out.datascout_report
    //)
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
