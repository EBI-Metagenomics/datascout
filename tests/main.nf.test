nextflow_pipeline {

    name "Test complete datascout pipeline"
    script "main.nf"

    test("Should run datascout pipeline end-to-end with test data") {

        setup {
            // Create mini NCBI database for testing
            "${projectDir}/tests/reference_databases/ncbi/make_mini_db.py".execute().waitFor()
        }

        when {
            params {
                // Use the example samplesheet from assets
                samplesheet = "${projectDir}/assets/samplesheet.csv"
                
                // Use the mini test databases created by setup
                taxdump = "${projectDir}/tests/reference_databases/ncbi/mini_taxdump.tar.gz"
                sqlite = "${projectDir}/tests/reference_databases/ncbi/mini_trypanosoma.sqlite"
                
                // Set small for fast test
                max_runs = 1
                max_orthodb_clusters = 1
                sourmash = false // sourmash will take too long
                
                // Use test output directory
                outdir = "$outputDir"
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert path("$outputDir").exists() },
                
                // Check pipeline_info directory and files
                { assert path("$outputDir/pipeline_info").exists() },
                { assert path("$outputDir/pipeline_info/software_versions.yml").exists() },
                
                { assert path("$outputDir").list().any { it.fileName.toString().endsWith("_reheaded_assembly.fasta") } },
                { assert path("$outputDir").list().any { it.fileName.toString().endsWith("_ENA_filtered_rna.csv") } },
                { assert path("$outputDir").list().any { it.fileName.toString().endsWith("_default_rna_fastq_dir") } },
                { assert path("$outputDir/test_sample1_default_rna_fastq_dir").list().size() == 2 },
                { assert path("$outputDir/test_sample1_tax_ranks.tsv").exists() },
                
                { assert path("$outputDir/test_sample1_orthodb_dir").exists() },
                { assert path("$outputDir/test_sample1_orthodb_dir").list().size() > 0 },
                
                { assert path("$outputDir/test_sample1_uniprot_dir").exists() },
                { assert path("$outputDir/test_sample1_uniprot_dir").list().any { it.fileName.toString().endsWith("_uniprot_proteins.faa") } },
                { assert path("$outputDir/test_sample1_uniprot_dir").list().any { it.fileName.toString().endsWith("_uniprot_raw.faa") } },
                
                { assert path("$outputDir/test_sample1_rfam_dir").exists() },
                { assert path("$outputDir/test_sample1_rfam_dir/rfam_ids.txt").exists() }
            )
        }
    }
}