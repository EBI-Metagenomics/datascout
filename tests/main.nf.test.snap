{
    "Should run datascout pipeline end-to-end with test data": {
        "content": [
            8,
            [
                "GCA_003719455.1_ENA_filtered_rna.csv",
                "GCA_003719455.1_default_rna_fastq_dir",
                "GCA_003719455.1_default_rna_fastq_dir/SRR25010087_1.fastq",
                "GCA_003719455.1_default_rna_fastq_dir/SRR25010087_2.fastq",
                "GCA_003719455.1_reheaded_assembly.fasta",
                "pipeline_info",
                "test_sample1_orthodb_dir",
                "test_sample1_orthodb_dir/5690_sequences",
                "test_sample1_orthodb_dir/5690_sequences/combined_orthodb_5690.faa",
                "test_sample1_rfam_dir",
                "test_sample1_rfam_dir/rfam_ids.txt",
                "test_sample1_tax_ranks.tsv",
                "test_sample1_uniprot_dir",
                "test_sample1_uniprot_dir/5693_uniprot_proteins.faa",
                "test_sample1_uniprot_dir/5693_uniprot_raw.faa",
                "versions.yml"
            ],
            [
                "GCA_003719455.1_ENA_filtered_rna.csv:md5,15cad058a4845db7b543f89e6bb52ebb",
                "SRR25010087_1.fastq:md5,717e8d1c6558ef61a60b08296f4c8b3a",
                "SRR25010087_2.fastq:md5,c53cfbdc10ea4bde941845452ab5c2b9",
                "GCA_003719455.1_reheaded_assembly.fasta:md5,fa4c17141ca1ffcae42f9e181946d3f4",
                "combined_orthodb_5690.faa:md5,991dcb1c20509697a7c971e259302e39",
                "test_sample1_tax_ranks.tsv:md5,b06eebb92f06f93c2894fb5f4e685d13",
                "5693_uniprot_proteins.faa:md5,79ab68a535498bbee3ecd193487e3fc6",
                "5693_uniprot_raw.faa:md5,380a59819358217b2f99a6538ba4baa4",
                "versions.yml:md5,a91cee1595a147f0f330b15cc937f43a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-09-17T16:10:32.865209"
    }
}